infrastructure/src/main/java/org/corfudb/infrastructure/compaction/PollingTask.java

```diff
@@ -0,0 +1,259 @@
+package org.corfudb.infrastructure.compaction;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.metrics.micrometer.MicroMeterUtils;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.CheckpointTaskMsg;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.CheckpointTaskStatus;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.PollingTaskTriggerKeyMsg;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.PollingTaskTriggerMsg;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.StatusKeyMsg;
+import org.corfudb.infrastructure.compaction.proto.LogCompaction.StatusMsg;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.collections.CorfuQueue;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.CorfuStreamEntries;
+import org.corfudb.runtime.collections.StreamListener;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.RetryExhaustedException;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.corfudb.infrastructure.compaction.CompactionManager.COMPACTION_NAMESPACE;
+import static org.corfudb.infrastructure.compaction.CompactionManager.COMPACTION_STREAM_TAG;
+import static org.corfudb.infrastructure.compaction.CompactionManager.TRIGGER_TABLE_NAME;
+
+@Slf4j
+public class PollingTask implements Runnable, StreamListener {
+    public static final int STATUS_RETRY_NUM = 5;
+    public static final Duration STATUS_RETRY_MAX_DURATION = Duration.ofSeconds(3);
+    public static final String EXCEPTION_MESSAGE = "Task Status failed to report";
+
+    private final CorfuRuntime corfuRuntime;
+
+    private final CorfuStore corfuStore;
+
+    private final Table<StatusKeyMsg, StatusMsg, Message> statusTable;
+
+    private final CorfuQueue checkpointTaskQueue;
+
+    private final CheckpointTaskProcessor taskProcessor;
+
+    // The Thread pool for executing polling tasks.
+    private final ScheduledExecutorService pollingExecutor;
+
+    private volatile SafeSnapshot safeSnapshot;
+
+    private volatile String compactionTaskId;
+
+    public PollingTask(CorfuRuntime runtime, CorfuStore corfuStore, String diskModePath,
```

Comment:
I think it might be better to take the pollingExecutor out of this PollingTask class. A runnable should be decoupled from the executor which is supposed to execute the runnable. In this way the polling task can focus on only the task logics itself and stay away from the execution logics. We can have one class which implements StreamListener and holds executor, and another class which implements Runnable and holds the task logics. What do you think?